### ISSUE TITLE
conv2d and conv1d ,weight grad may be wrong

### ISSUE THREAD

**Issue by laizewei** [pytorch/pytorch#15501]
pytorch 1.0 windows/python3.6/cpu

```python
import torch 
import torch.nn.functional as F
a=[[[7.,3.]],[[3.,2.]]]
b=[[[3.,6.]],[[4.,2.]]]
aa = torch.tensor(a,requires_grad=True)
bb = torch.tensor(b,requires_grad=True)
dd =F.conv1d(aa,bb,padding=1,stride=2,dilation=2)
z=torch.sum(dd)
z.backward()
print('dd',dd,dd.size())
print(bb.grad)
#dd=[[[18.],[6.]],[[12.]],[4.]]]
#dd.grad=[[[1.],[1.]],[[1.],[1.]]]
#padding dd.grad=k 
k=[[[0.,1.,0.],[0.,1.,0.]],[[0.,1.,0.],[0.,1.,0.]]]
kk = torch.tensor(k,requires_grad=True)
output=F.conv_transpose1d(kk,aa,padding=2,stride=2,dilation=2,output_padding=0)
print(output)
#because stride=2,so  take part of output [[[0.,5.]],[[0.,5.]]] ,thas is bb.grad
```

```
dd tensor([[[18.],
         [ 6.]],

        [[12.],
         [ 4.]]], grad_fn=<SqueezeBackward1>) torch.Size([2, 2, 1])
tensor([[[0., 5.]],

        [[0., 5.]]])
tensor([[[10.,  0.,  5.]],

        [[10.,  0.,  5.]]], grad_fn=<SqueezeBackward1>)
```

This method can also be evaluated `aa.grad`.

```
import torch 
import torch.nn.functional as F
a=[[[7.,3.]],[[3.,2.]]]
b=[[[3.,6.]],[[4.,2.]]]
aa = torch.tensor(a,requires_grad=True)
bb = torch.tensor(b,requires_grad=True)
dd =F.conv1d(aa,bb,padding=1,stride=2,dilation=2)
z=torch.sum(dd)
z.backward()
print('dd',dd,dd.size())
print(aa.grad)
#dd=[[[18.],[6.]],[[12.]],[4.]]]
#dd.grad=[[[1.],[1.]],[[1.],[1.]]]
#padding dd.grad=k 
k=[[[0.,1.,0.],[0.,1.,0.]],[[0.,1.,0.],[0.,1.,0.]]]
kk = torch.tensor(k,requires_grad=True)
output=F.conv_transpose1d(kk,bb,padding=2,stride=2,dilation=2,output_padding=0)
print(output)
#because stride=2,so  take part of output [[[0.,8.]],[[0.,8.]]] ,thas is aa.grad.
```

```
dd tensor([[[18.],
         [ 6.]],

        [[12.],
         [ 4.]]], grad_fn=<SqueezeBackward1>) torch.Size([2, 2, 1])
tensor([[[0., 8.]],

        [[0., 8.]]])
tensor([[[7., 0., 8.]],

        [[7., 0., 8.]]], grad_fn=<SqueezeBackward1>)
```

**question:**

when padding=stride=dalition,aa.grad is correct，but bb.grad is wrong?

```python
import torch 
import torch.nn.functional as F
a=[[[7.,3.]],[[3.,2.]]]
b=[[[3.,6.]],[[4.,2.]]]
aa = torch.tensor(a,requires_grad=True)
bb = torch.tensor(b,requires_grad=True)
dd =F.conv1d(aa,bb,padding=2,stride=2,dilation=2)
z=torch.sum(dd)
z.backward()
print('dd',dd,dd.size())
print(aa.grad)
#dd=[[[42.,21.],[14.,28.]],[[18.,9.,],[6.,12.]]]
#dd.grad=[[[1.,1.],[1.,1.]],[[1.,1.],[1.,1.]]]
#padding dd.grad=k 
k=[[[0.,1.,1.,0.],[0.,1.,1.,0.]],[[0.,1.,1.,0.],[0.,1.,1.,0.]]]
kk = torch.tensor(k,requires_grad=True)
output=F.conv_transpose1d(kk,bb,padding=3,stride=2,dilation=2,output_padding=0)
print(output)
#because stride=2,so  take part of output [[[15.,0.]],[[15.,0.]]] ,thas is aa.grad
```

```python
import torch 
import torch.nn.functional as F
a=[[[7.,3.]],[[3.,2.]]]
b=[[[3.,6.]],[[4.,2.]]]
aa = torch.tensor(a,requires_grad=True)
bb = torch.tensor(b,requires_grad=True)
dd =F.conv1d(aa,bb,padding=2,stride=2,dilation=2)
z=torch.sum(dd)
z.backward()
print('dd',dd,dd.size())
print(bb.grad)
#dd=[[[42.,21.],[14.,28.]],[[18.,9.,],[6.,12.]]]
#dd.grad=[[[1.,1.],[1.,1.]],[[1.,1.],[1.,1.]]]
#padding dd.grad=k 
k=[[[0.,1.,1.,0.],[0.,1.,1.,0.]],[[0.,1.,1.,0.],[0.,1.,1.,0.]]]
kk = torch.tensor(k,requires_grad=True)
output=F.conv_transpose1d(kk,aa,padding=3,stride=2,dilation=2,output_padding=0)
print(output)
#because stride=2,so  take part of output [[[15.,0.]],[[15.,0.]]] ,but is not  bb.grad.
```

```
dd tensor([[[42., 21.],
         [14., 28.]],

        [[18.,  9.],
         [ 6., 12.]]], grad_fn=<SqueezeBackward1>) torch.Size([2, 2, 2])
tensor([[[10., 10.]],

        [[10., 10.]]])
tensor([[[ 0., 15.,  0.]],

        [[ 0., 15.,  0.]]], grad_fn=<SqueezeBackward1>)
```

Was it my method that did the calculation wrong?

**Comment by laizewei** [comment]
now I use col2im to calculate the derivative，pytorch is true.

**Comment by soumith** [comment]
I dont see the incorrectness.

We'll start with `aa`. After padding = 1, it effectively becomes:

batch0 = [0, 7, 3, 0]
batch 1 = [0, 3, 2, 0]

And then `bb`, after `dilation=2`, it effectively becomes something like:

oc0 = 3, 0, 6, 0
oc1 = 4, 0, 2, 0

The output will be:

location 0, 0:

`[0, 7, **3**, 0] * [3, 0, **6**, 0]` = 18

location 0, 1:

`[0, 7, **3**, 0] * [4, 0, **2**, 0]` = 6

**Comment by soumith** [comment]
I dont see the incorrectness.

We'll start with `aa`. After padding = 1, it effectively becomes:

batch0 = [0, 7, 3, 0]
batch 1 = [0, 3, 2, 0]

And then `bb`, after `dilation=2`, it effectively becomes something like:

oc0 = 3, 0, 6, 0
oc1 = 4, 0, 2, 0

The output will be:

location 0, 0:

`[0, 7, **3**, 0] * [3, 0, **6**, 0]` = 18

location 0, 1:

`[0, 7, **3**, 0] * [4, 0, **2**, 0]` = 6

Similarly, location 1, 0 and 1, 1 will be:
`[0, 3, **2**, 0] * [3, 0, **6**, 0]` = 12
`[0, 3, **2**, 0] * [4, 0, **2**, 0]` = 4

When you calculate `aa`'s grad, it'll be grad_output (which is just ones) times corresponding terms in `bb`.
So, `aa`'s grad_output will come out to: 6 * 1 + 2 * 1 = 8. (the first column is 0, because it didn't contribute to output)

Similarly, `bb`'s grad_output will come out to: 3 * 1 + 2 * 1 = 5. (again, bb's first column is 0, because it didn't contribute to output)
